src/memo/memo.ml

```diff
@@ -337,11 +337,122 @@ module Deps_so_far = struct
     | Compute_started { deps_reversed } -> deps_reversed
 end
 
+module Cache_lookup = struct
+  (* Looking up a value cached in a previous run can fail in three possible
+     ways:
+
+     - [Not_found]: either the value has never been computed before, or the last
+     computation attempt failed.
+
+     - [Out_of_date]: we found a value computed in a previous run but it is out
+     of date because one of its dependencies changed; we return the old value so
+     that it can be compared with a new one to support the early cutoff.
+
+     - [Cancelled _]: the cache lookup has been cancelled because of a
+     dependency cycle. This outcome indicates that a dependency cycle has been
+     introduced in the current run. If a cycle existed in a previous run, the
+     outcome would have been [Not_found] instead. *)
+  module Failure = struct
+    type 'a t =
+      | Not_found
+      | Out_of_date of 'a
+      | Cancelled of { dependency_cycle : Cycle_error.t }
+  end
+
+  module Result = struct
+    type 'a t = ('a, 'a Failure.t) result
+  end
+end
+
+(* A fiber that can be shared but is still computed at most once. An equivalent
+   of ['a Lazy.t] but for asynchronous computations. *)
+module Once = struct
+  type 'a state =
+    | Not_forced of (unit -> 'a Fiber.t)
+    | Forced of 'a Fiber.Ivar.t
+
+  type 'a t = { mutable state : 'a state }
+
+  let create thunk = { state = Not_forced thunk }
+
+  let force t =
+    match t.state with
+    | Forced ivar ->
+      let+ res = Fiber.Ivar.read ivar in
+      res
+    | Not_forced thunk ->
+      let ivar = Fiber.Ivar.create () in
+      t.state <- Forced ivar;
+      let* result = thunk () in
+      let+ () = Fiber.Ivar.fill ivar result in
+      result
+
+  (* Almost a monadic bind but [f] returns a [Fiber.t], not a [t]. *)
+  let and_then t ~f = create (fun () -> Fiber.bind (force t) ~f)
+end
+
+(* An attempt to sample the current value of a node. It's an "attempt" because
```

Comment:
Yep, will rephrase it.

Reply:
Fine. I still disagree with the "because", but at least the number of reasons is now less than 2. :-)
(if you need a historical reminder, the reason it's called attempt is because we needed a word to distinguish between the act of sampling and the result of sampling and this was our best *attempt* :-D)